[GStreamer][EME] imported/w3c/web-platform-tests/encrypted-media/clearkey-mp4-playback-temporary-clear-encrypted-segmented.https.html is failing

philn opened the pull request:
#### 35dba6c294cd55aaf82e1037f5f69b55808a91dc
<pre>
[GStreamer][EME] imported/w3c/web-platform-tests/encrypted-media/clearkey-mp4-playback-temporary-clear-encrypted-segmented.https.html is failing
<a href="https://bugs.webkit.org/show_bug.cgi?id=211375">https://bugs.webkit.org/show_bug.cgi?id=211375</a>

Reviewed by Xabier Rodriguez-Calvar.

Wrap the decryptor in a new bin along with parsebin. This should allow seamless switching between
encrypted and clear content processing. As this new parser is auto-plugged by parsebin in
urisourcebin we have to be cautious regarding the infinite auto-plugging outcome by removing our new
parser from the autoplug-factories passed to the inner parsebin element.

Canonical link: <a href="https://commits.webkit.org/295500@main">https://commits.webkit.org/295500@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d5f665ccbf252c476c6095cefab383cd97fa3b46

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/105263 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/24975 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/15400 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/110474 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/55919 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/107304 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/25407 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/33518 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/79953 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/108269 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/19812 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/95009 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/60259 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/104741 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/19567 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/13083 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/55317 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/89266 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/13128 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/113078 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/32421 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/23886 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/89026 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/32785 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/91225 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/88664 "Build is in progress. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; run-api-tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22611 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/33559 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/11347 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/27845 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/32345 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/37757 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/32123 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/35468 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/33692 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

calvaris commented on Source/WebCore/platform/graphics/gstreamer/eme/WebKitThunderParser.cpp:
```diff
@@ -0,0 +1,171 @@
+/* GStreamer Thunder Parser
+ *
+ * Copyright (C) 2025 Metrological
```
Should be Comcast Inc.

calvaris commented on Source/WebCore/platform/graphics/gstreamer/eme/WebKitThunderParser.cpp:
```diff
@@ -0,0 +1,171 @@
+        ALLOW_DEPRECATED_DECLARATIONS_BEGIN;
+        GValueArray* result;
+
+        GUniquePtr<GList> factories(gst_element_factory_list_get_elements(GST_ELEMENT_FACTORY_TYPE_DECODABLE, GST_RANK_MARGINAL));
```
I think you have to call `gst_plugin_feature_list_free` or you'd be leaking the factories, according to doc.

A more complicated but maybe nicer would be tweaking the implementation of GUniquePtr to be able to make a quick inline declaration of the deleter.

philn replied:
I'm not sure how to do that yet, left as potential follow-up!

calvaris commented on Source/WebCore/platform/graphics/gstreamer/eme/WebKitThunderParser.cpp:
```diff
@@ -0,0 +1,171 @@
+    }), self);
+
+    g_signal_connect(self->priv->parser.get(), "pad-added", G_CALLBACK(+[](GstElement*, GstPad* pad, gpointer userData) {
+        static uint64_t counter = 0;
```
Do we need 64 bits for this? Are planning to have so many pads? :smile:

calvaris commented on Source/WebCore/platform/graphics/gstreamer/eme/WebKitThunderParser.h:
```diff
@@ -0,0 +1,37 @@
+/* GStreamer Thunder Parser
+ *
+ * Copyright (C) 2025 Metrological
```
comcast

philn committed:
[GStreamer][EME] imported/w3c/web-platform-tests/encrypted-media/clearkey-mp4-playback-temporary-clear-encrypted-segmented.https.html is failing
https://bugs.webkit.org/show_bug.cgi?id=211375

Reviewed by Xabier Rodriguez-Calvar.

Wrap the decryptor in a new bin along with parsebin. This should allow seamless switching between
encrypted and clear content processing. As this new parser is auto-plugged by parsebin in
urisourcebin we have to be cautious regarding the infinite auto-plugging outcome by removing our new
parser from the autoplug-factories passed to the inner parsebin element.

Canonical link: https://commits.webkit.org/295500@main
